src/corehost/cli/fxr/fx_muxer.cpp

```diff
@@ -68,6 +88,15 @@ static int execute_app(
     if (code != StatusCode::Success)
         return code;
 
+    {
+        // Track an empty 'active' context so that host context-based APIs can work properly when
+        // the runtime is loaded through non-host context-based APIs.
+        std::lock_guard<std::mutex> lock{ g_context_lock };
+        assert(g_active_host_context == nullptr);
+        g_active_host_context.reset(new host_context_t());
+        g_context_initializing.store(false);
```

Comment:
Yeah... I currently just have any calls to get properties for the active context error out if the active context is created through this path (i.e. empty). I guess another option would be to just have the config properties (since the old `hostpolicy` interfaces don't expose the calculated properties), but that seems confusing as well.

Would this be the race you are talking about:
- thread (1) calls `hostfxr_main*`
- thread (2) calls `hostfxr_initialize_for_runtime_config` 
- (1) `hostfxr_main*` marks the context as initializing
- (2) `hostfxr_initialize_for_runtime_config` waits
- (1) creates an 'empty' context, signals that initialization is done (but runtime is not actually loaded)
- (2) unblocks, creates a secondary context, and calls `hostfxr_get_runtime_delegate`, which calls into `hostpolicy`
- (2) call fails because runtime has not been loaded yet

Reply:
I can't think of a way to address that properly going through the old `hostpolicy` APIs...

I guess if we did the riskier option of making `hostfxr` try the new APIs first, then fall back to the old ones, it would indirectly block using a mix of old/new `hostpolicy` APIs), so it would avoid it?

Reply:
I was trying to go through this in my head a bit more. The two options we discussed:
1. Add a field (e.g. a callback) to `host_interface_t` that gets stored by `corehost_load` and used as a way to notify hostfxr when the runtime is actually loaded and pass over any other information (e.g. properties)
2. Use the new hostpolicy APIs if available. Use the old ones as a fallback. This would effectively avoid using a mix of old/new APIs.

- Both are a significant change in how the existing hostfxr entry points would function. 1 is maybe a bit less risk?
- 1 would introduce something for this specific scenario (old/new API mix) that we'd basically be stuck with forever.
- I would argue that 2 is cleaner and is what would make sense long-term, but maybe not that great so late in the release.
- With the way I have hostpolicy right now, I think the main difference between the logic in the old and new APIs is how the app args are determined: the new function for running an app just expect argc/argv to be exactly what should be passed to the app, while `corehost_main` takes the input argc/argv and figures out the app argc/argv based on the host mode

Maybe a third option would be to make hostpolicy block waiting for the runtime to actually load (instead of just failing) if it hits this situation? Something like:
- hostfxr passes nullptr for the hostpolicy context handle when it needs to perform an operation via the new APIs but 'active' context on hostfxr was created via the old APIs. Any other situation should pass a proper handle.
- If hostpolicy gets nullptr for the context handle (for operations that allow it - e.g. getting a delegate), it uses the global hostpolicy_context. If there is no tracked context (runtime has not been loaded and initialized yet), it blocks until it is loaded (or a load attempt fails).
- Failure on initializing the runtime would be kind of weird in that everything is actually dead once initializing coreclr fails, but the secondary context was successfully initialized since it was before we tried/failed (whereas going through only the new APIs, after the first coreclr initialization fails, future attempts to initialize a context would fail)
- It would mean that a null context handle has a special / hidden meaning for some hostpolicy operations. (And someone directly using hostpolicy could certainly get into a bad spot without knowing the special-case)
- It should only affect this specific scenario of the concurrent use of old/new APIs (shouldn't be hit if just using old APIs, just using new APIs, or using old/new APIs synchronously).

Not exactly great either, but it would at least be pretty targeted?

Reply:
I tried adding an initialization option to hostpolicy to handle this in my latest commit. I think it addresses the synchronization problems with mixing old and new APIs without affecting other scenarios (getting runtime properties of the active context is still blocked if it was created through the old APIs though).

Reply:
We can do this in a subsequent change... but I think we should do it before releases. The scenario of running app through old APIs and then using hosting APIs to load component should be something which works fully (Even though right now I can't come up with a valid scenario where the properties would matter, since the most likely use case is COM/WinRT which won't need the properties).

Reply:
I like it. Created #6197 for doing this in a subsequent change.